sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,365 @@
+	}
+	defer func() {
+		if err != nil {
+			poll.Close()
+		}
+	}()
+	if err = poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener = &Listener{
+		reqs: make(chan *Request),
+		errs: make(chan error, 1),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+// Close stops the listener and closes the kernel communication file.
+// Returns once all waiting goroutines terminate and the communication file
+// is closed.
+func (l *Listener) Close() error {
+	l.tomb.Kill(nil)
+	origStatus := atomic.SwapUint32(&l.status, statusClosed)
+	switch origStatus {
+	case statusReady:
+		// A goroutine was never spawned with the listener tomb.
+		// Spawn one, so the tomb can die.
+		l.tomb.Go(func() error {
+			<-l.tomb.Dying()
+			return nil
+		})
+	case statusRunning:
+	case statusClosed:
+		return ErrListenerAlreadyClosed
+	}
+	// Stop the run loop from waiting on epoll events
+	err1 := l.poll.Close()
+	// Wait for all waiters to send back deny responses to the kernel
+	l.tomb.Wait()
+	err2 := l.notifyFile.Close()
+	if err1 != nil {
+		return err1
+	}
+	return err2
+}
+
+// Reqs returns a read-only channel through which requests may be received.
+func (l *Listener) Reqs() <-chan *Request {
+	return l.reqs
+}
+
+// Errs returns a read-only channel through which errors may be received.
+func (l *Listener) Errs() <-chan error {
+	return l.errs
+}
+
+// Run reads and dispatches kernel requests until the listener is closed.
+//
+// Run should only be called once per listener object. If called more than once,
+// Run returns an error. Otherwise, waits until the listener stops, and returns
+// the cause as an error. If the listener was intentionally stopped via the
+// Close() method, returns nil.
+func (l *Listener) Run() error {
+	if !atomic.CompareAndSwapUint32(&l.status, statusReady, statusRunning) {
+		currStatus := atomic.LoadUint32(&l.status)
+		switch currStatus {
+		case statusRunning:
+			return ErrListenerAlreadyRun
+		case statusClosed:
+			return ErrListenerAlreadyClosed
+		default:
+			return fmt.Errorf("listener has unexpected status: %d", currStatus)
+		}
+	}
+	// This is the first and only time calling Run().
+	// Even if Close() kills the tomb before l.tomb.Go() occurs, a panic will
+	// not occur, since this new goroutine will (immediately after l.tomb.Err()
+	// is called) return and close the tomb's dead channel, as it is the last
+	// and only tracked goroutine to terminate.
+	l.tomb.Go(func() error {
+		var err error
+		for {
+			err = l.tomb.Err()
+			if err != tomb.ErrStillAlive {
+				break
+			}
+			err = l.runOnce()
+			if err != nil {
+				if err == ErrListenerAlreadyClosed {
+					// Close() was called externally.
+					// Do not treat this as an error.
+					err = nil
+				}
+				break
+			}
+		}
+		if err != nil {
+			l.errs <- err
+		}
```

Comment:
we stop on the first error here, it's a bit unclear to me why we need a channel for errors? is it to help stop things listening on Reqs?  shouldn't we expose a Dying channel instead?

Reply:
Great idea. I've replaced the `errs` and `Errs()` with a new method, `Err()`, which exports `l.tomb.Wait()`. I think it's best if reading from `Err()` blocks until all tracked goroutines have terminated (after which time the listener socket closes) so that the caller could immediately re-register a new listener afterwards.